patches/api/0246-Registry-Modification-API.patch

```diff
@@ -0,0 +1,703 @@
+new file mode 100644
+index 0000000000000000000000000000000000000000..4cf02f2b496a3c89b8d071a16147677b41eb5b28
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/registry/ReferenceImpl.java
+@@ -0,0 +1,52 @@
++package io.papermc.paper.registry;
++
++import com.google.common.base.Suppliers;
++import java.util.function.Supplier;
++import net.kyori.adventure.key.Key;
++import org.bukkit.Keyed;
++import org.bukkit.NamespacedKey;
++import org.bukkit.Registry;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++import java.util.NoSuchElementException;
++
++record ReferenceImpl<T extends Keyed>(@NotNull Supplier<Registry<T>> registry, @NotNull Key key) implements Reference<T> {
++
++    ReferenceImpl(@NotNull Registry<T> registry, @NotNull Key key) {
++        this(() -> registry, key);
++    }
++
++    ReferenceImpl(@NotNull RegistryKey<T> registryKey, @NotNull Key key) {
++        this(Suppliers.memoize(() -> RegistryAccess.INSTANCE.getRegistry(registryKey)), key);
++    }
++
++    @Override
++    public @NotNull T value() {
++        final @Nullable T value = this.registry.get().get(this.key);
++        if (value == null) {
++            throw new NoSuchElementException("No such value with key " + this.key);
++        }
++        return value;
++    }
++
++    @Override
++    public @Nullable T valueOrNull() {
++        return this.registry.get().get(this.key);
++    }
++
++    @Override
++    public boolean exists() {
++        return this.registry.get().get(this.key) != null; // TODO replace with Registry#containsKey() when that's added
++    }
++
++    @Override
++    public @NotNull Key key() {
++        return this.key;
++    }
++
++    @Override
++    public @NotNull NamespacedKey getKey() {
++        return this.key instanceof NamespacedKey nsKey ? nsKey : new NamespacedKey(this.key.namespace(), this.key.value());
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/registry/RegistryAccess.java b/src/main/java/io/papermc/paper/registry/RegistryAccess.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..d998e607486a5f19b0e9f6d139bb3babf7e66d20
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/registry/RegistryAccess.java
+@@ -0,0 +1,42 @@
++package io.papermc.paper.registry;
++
++import java.util.ServiceLoader;
++import org.bukkit.Keyed;
++import org.bukkit.Registry;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++/**
++ * The {@link #INSTANCE} provides access to registries given a {@link RegistryKey}.
++ * <p>
++ * Replacement for {@link org.bukkit.Bukkit#getRegistry(Class)}.
++ */
++@ApiStatus.NonExtendable
++@ApiStatus.Experimental
++public interface RegistryAccess {
++
++    RegistryAccess INSTANCE = ServiceLoader.load(RegistryAccess.class).iterator().next();
```

Comment:
There's a nice Services.load() method that adventure offers so you don't have to use any iterator funnies.

Additionally, are we sure we want this as a publicly exposed field instead of an access method?

Reply:
Yeah, we can't use adventures. I already tried it. You can't change the classloader, and the classloader it uses will fail for tests, you need the thread context classloader for it to load the service for tests

Reply:
> are we sure we want this as a publicly exposed field instead of an access method?

yeah, idc either way. It just can't be on the Bukkit singleton, or well I guess it can, just not part of the Server interface.